### PR TITLE
Speech workloads mask fix

### PR DESCRIPTION
**Pull request by sourabh2k15**
in speech workload cases mask is already given by target_paddings , it's a 2d array containing paddings for individual examples as well as the batch dimension

### CODE REVIEW

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py [hunk 1]
```diff
@@ -166,10 +166,11 @@ def loss_fn(
                                        logit_paddings,
                                        targets,
                                        target_paddings)
-    # mask_batch is assumed to be shape [batch].
+    # mask_batch is assumed to be shape [batch] for other workloads
+    # but in speech cases it's [batch, target_len].
     if mask_batch is not None:
-      per_example_losses *= mask_batch
       mask_batch = jnp.logical_and(mask_batch, 1 - target_paddings)
```
Comment (runame) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py:172]: I think we multiplied by `mask_batch` after this line because `target_paddings` are already included in the ctc loss calculation, but I think it should not make a difference?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py [hunk 2]
```diff
@@ -166,10 +166,11 @@ def loss_fn(
...
+      per_example_losses *= mask_batch[:, 0]
```
Comment (runame) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py:173]: Do we expect the mask to be the same across the second dimension (of size target_len) or why is it correct to only use the first slice of the second dimension?
If the size of `mask_batch` is (batch_size, target_len), can't we just also pass it together with target_paddings to `self.ctc_loss`? If we just want to provide the user the option to mask out the contribution of specific examples to the loss, the shape of `mask_batch` should maybe just be (batch_size,).

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py [hunk 3]
```diff
@@ -166,10 +166,11 @@ def loss_fn(
...
     else:
       mask_batch = 1 - target_paddings
     n_valid_examples = jnp.maximum(mask_batch.sum(), 1)
```
Comment (runame) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py:176]: If the shape of `mask_batch` is (batch_size, target_len), do we really want to sum over both dimensions to get `n_valid_examples` or should it just be the number of examples for which mask batch is not all zero?

**Review by runame — COMMENTED** [review]
Some more questions/comments:
1. This logic did not change in any recent refactor AFAIK, so I'm confused why it only know starts to make problems. Maybe we just didn't run the speech workloads since we introduced this logic? I guess the tests didn't fail because they do not provide an extra mask?
2. Is the PyTorch version of the workload not also affected?
3. ImageNet ViT is also affected by this (see #331) and it worked fine before. I suspect the issue is caused by something non-specific to any individual workload and should be fixed on that level.

**Comment by runame** [comment]
I think I have found a potential cause of the bug, [this commit](https://github.com/mlcommons/algorithmic-efficiency/commit/8d3fd81d9e5fcecb0435fee6faa9ff591f5460d3) creates weights by default and not only when they are needed due to padding being applied. [This line](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/data_utils.py#L41) will create a 2d mask even for ImageNet for which this is not desired.

@pomonam I guess you wanted to be able to always broadcast the weights when using a shared input pipeline? I guess we can still achieve this with a few changes.

**Comment by pomonam** [comment]
@runame Thank you for debugging this! At the time of the commit, I don't think `mask_shape = targets_shape if len(targets_shape) < 3 else targets_shape[0]` was triggered, since I remember running imagenet workloads. But, as you suggested, an easy fix here for now would be to revert these lines (indent them again).

Yes, I wanted to broadcast the weights when using the shared input pipeline (also for imagenet-v2) and I also agree that we only need mask during eval. I can make changes on other places to support this, so how about we go ahead revert my commits and then I will change workloads (e.g., mnist) that relied on my commit?

**Comment by pomonam** [comment]
#349 should consume this PR.
